flink-staging/flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkShell.scala

```diff
@@ -0,0 +1,148 @@
+import org.apache.flink.runtime.minicluster.LocalFlinkMiniCluster
+
+import scala.tools.nsc.interpreter._
+
+
+object FlinkShell {
+
+  val LOCAL = 0;
+  val REMOTE = 1;
+  val UNDEFINED = -1;
+
+  var bufferedReader: BufferedReader = null;
+
+  def main(args: Array[String]) {
+
+    // scopt, command line arguments
+    case class Config(
+        port: Int = -1,
+        host: String = "none",
+        externalJars: Option[Array[String]] = None,
+        flinkShellExecutionMode : Int = UNDEFINED)
+
+    val parser = new scopt.OptionParser[Config]("start-scala-shell.sh") {
+      head ("Flink Scala Shell")
+
+      cmd("local") action {
+        (_,c) => c.copy( host = "none", port = -1, flinkShellExecutionMode = LOCAL)
+      } text("starts Flink scala shell with a local Flink cluster\n") children(
+        opt[(String)] ("addclasspath") abbr("a") valueName("<path/to/jar>") action {
+          case (x,c) =>
+            val xArray = x.split(":")
+            c.copy(externalJars = Option(xArray))
+          } text("specifies additional jars to be used in Flink\n")
+        )
+
+      cmd("remote") action { (_, c) =>
+        c.copy(flinkShellExecutionMode = REMOTE)
+      } text("starts Flink scala shell connecting to a remote cluster\n") children(
+        arg[String]("<host>") action { (h, c) =>
+          c.copy(host = h) }
+          text("remote host name as string"),
+        arg[Int]("<port>") action { (p, c) =>
+          c.copy(port = p) }
+          text("remote port as integer\n"),
+        opt[(String)] ("addclasspath") abbr("a")  valueName("<path/to/jar>") action {
+          case (x,c) =>
+            val xArray = x.split(":")
+            c.copy(externalJars = Option(xArray))
+          } text("specifies additional jars to be used in Flink")
+
+      )
+      help("help") abbr("h") text("prints this usage text\n")
+    }
+
+
+    // parse arguments
+    parser.parse (args, Config () ) match {
+      case Some(config) =>
+        startShell(config.host,
+          config.port,
+          config.flinkShellExecutionMode,
+          config.externalJars)
+
+      case _ => println("Could not parse program arguments")
+    }
+  }
+
+
+  def startShell(
+      userHost : String, 
+      userPort : Int,
+      executionMode : Int,
+      externalJars : Option[Array[String]] = None): Unit ={
+    
+    println("Starting Flink Shell:")
+
+    var cluster: LocalFlinkMiniCluster = null
+
+    // either port or userhost not specified by user, create new minicluster
+    val (host,port) = if (executionMode == LOCAL) {
+      cluster = new LocalFlinkMiniCluster(new Configuration, false)
+      cluster.start()
+      val port = cluster.getLeaderRPCPort
+      println(s"\nStarting local Flink cluster (host: localhost, port: $port).\n")
+      ("localhost",port)
+    } else if(executionMode == UNDEFINED) {
+      println("Error: please specify execution mode:")
+      println("[local | remote <host> <port>]")
+      return
```

Comment:
Return statements are breaking the control flow. This is not a good style. Would be great to get rid of them.